docs/design/structs-interfaces.md

```diff
@@ -0,0 +1,304 @@
+}
+
+type ProviderWithProviderMeta interface {
+  GetMetaSchema(ctx context.Context) (*tfprotov5.Schema, []*tfprotov5.Diagnostic)
+}
+
+type ResourceFactory interface {
+  NewResource(Provider) (Resource, []*tfprotov5.Diagnostic)
+}
+
+type Resource interface {
+  GetSchema(context.Context) (*tfprotov5.Schema, []*tfprotov5.Diagnostic)
+  Create(context.Context, ResourceCreateRequest, ResourceCreateResponse)
+  Read(context.Context, ResourceReadRequest, ResourceReadResponse)
+  Update(context.Context, ResourceUpdateRequest, ResourceUpdateResponse)
+  Delete(context.Context, ResourceDeleteRequest, ResourceDeleteResponse)
+  Import(context.Context, ResourceImportRequest, ResourceImportResponse)
+}
+
+type Importer interface {
+  Import(context.Context, string) ([]*tfprotov5.ImportedResource, []*tfprotov5.Diagnostic)
+}
+
+type Validater interface {
+  Validate(context.Context, tftypes.Value) []*tfprotov5.Diagnostic
+}
+
+type CustomPlanner interface {
+  ModifyPlan(ctx context.Context, priorState, proposedState, config tftypes.Value) (tftypes.Value, []*tfprotov5.Diagnostic)
+}
+```
+
+### Alternative 3: Mix of structs and interfaces
+
+Example framework implementation:
+
+```go
+type Provider struct {
+	Resources     map[string]Resource
+	ConfigureFunc func(context.Context, *tfprotov5.ConfigureProviderRequest, *tfprotov5.ConfigureProviderResponse, *ProviderState)
+}
+
+type Resource struct {
+	Schema                *tfprotov5.Schema
+	Implementation        func(p ProviderState) ResourceImplementation
+}
+
+// ResourceImplementation defines the methods of this resource.
+type ResourceImplementation interface {
+	Create(context.Context) []*tfprotov5.Diagnostic
+	Read(context.Context) []*tfprotov5.Diagnostic
+	Update(context.Context) []*tfprotov5.Diagnostic
+	Destroy(context.Context) []*tfprotov5.Diagnostic
+}
+```
+
+## Tradeoffs
+
+### Runtime state
+
+This section concerns stateful data generated during the lifecycle of the provider, not Terraform state.
+
+For example, consider the `meta` parameter in the CRUD functions of [`helper/schema#Resource`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#Resource), which is conventionally used by providers to store a configured API client. Since this data lives in a framework-owned type and not a provider-owned type, its type cannot be known at compile time and it is represented as `interface{}`. Its type must be asserted at the top of each CRUD method where it is used.
+
+#### Structs
+
+If providers, resources, and data sources are represented as structs in framework code, provider code will instantiate values of those structs. 
+
+Certain struct fields could be populated by the framework: for `Resource`, this could include `ProviderMeta` and `TerraformVersion`. Provider code cannot add additional struct fields to store strongly-typed data. This lack of flexibility is a **disadvantage**, although it likely also prevents some anti-patterns.

```

Comment:
```suggestion
Certain struct fields could be populated by the framework: for `Resource`, this could include the equivalent of SDKv2's `meta` value and `TerraformVersion`. Provider code cannot add additional struct fields to store strongly-typed data. This lack of flexibility is a **disadvantage**, although it likely also prevents some anti-patterns.
```

Just a suggestion. I think talking about `ProviderMeta` was a mistake on my part and I needlessly complicated everything, sorry.